### PR TITLE
skeleton/*: imported from php-ext

### PR DESCRIPTION
**Pull request by davidam**

### CODE REVIEW

**Comment by syohex** [comment]
Why do you import them ?

**Comment by trashofmasters** [comment]
I wonder if he didn't make a PR by accident, otherwise why would there be no description?

**Comment by ejmr** [comment]
There was probably no description because David discussed it with me on IRC.

@davidam I think you should give a brief description of the purpose and intent behind the pull request, so that everyone else can provide any feedback.

**Comment by davidam** [comment]
Skeletons are templates to fill fields and write less in php. You can call
any skeleton defined with M-x and try it.

Kind regards

2015-05-20 18:26 GMT+02:00 Eric James Michael Ritz <notifications@github.com
>:

> There was probably no description because David discussed it with me on
> IRC.
>
> @davidam <https://github.com/davidam> I think you should give a brief
> description of the purpose and intent behind the pull request, so that
> everyone else can provide any feedback.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ejmr/php-mode/pull/250#issuecomment-103947759>.
>



-- 
David Arroyo Menéndez
http://www.davidam.com

**Comment by ejmr** [comment]
Personally I think it will be nice to have these in PHP Mode for developers who use skeletons---[useful reading material](http://www.emacswiki.org/emacs/SkeletonMode).  I will add some information to the README describing what they provide and some basic steps for using them for any interested developers.

But I will merge this at the end of the week at the earliest, because I would like @syohex and @trashofmasters and anyone else interested to have a chance to provide feedback before then.

**Comment by trashofmasters** [comment]
I am a user of YaSnippet but I seldom use it outside of PHP-development
therefore I'd happily drop the dependency on favour of of the Emacs
built-in functionality. I'm generally happy to have the some Skeleton
functionality embedded into PHP-Mode.

Having said that, I reckon Skeletons as a fantastic way to reduce
boilerplate code (e.g. creating classes, traits, interfaces, test cases,
etc) but I am still unsure about having built-in PHP functions implemented
as skeletons.

To be completely fair I haven't had the time to test @davidam's code and
have never worked with Skeleton before. So I'd assume it prompts the user
for input using the echo-area/minibuffer. If that's the case then I'm not
completely sold on it.

When I code I personally find it very distracting when the minibuffer gets
the focus :) Fortunately I believe this can easily be worked around, for
example I'm pretty confident that somewhere someone has already written a
Helm interface to list Skeletons.

To recap: +1 for the work, although I would limit the use of skeletons for
"file-level" boilerplate.

**Comment by davidam** [comment]
Thanks for the comments. I'm happy with the plans to accept the work.

Regards.

2015-05-21 12:35 GMT+02:00 Andrea Turso <notifications@github.com>:

> I am a user of YaSnippet but I seldom use it outside of PHP-development
> therefore I'd happily drop the dependency on favour of of the Emacs
> built-in functionality. I'm generally happy to have the some Skeleton
> functionality embedded into PHP-Mode.
>
> Having said that, I reckon Skeletons as a fantastic way to reduce
> boilerplate code (e.g. creating classes, traits, interfaces, test cases,
> etc) but I am still unsure about having built-in PHP functions implemented
> as skeletons.
>
> To be completely fair I haven't had the time to test @davidam's code and
> have never worked with Skeleton before. So I'd assume it prompts the user
> for input using the echo-area/minibuffer. If that's the case then I'm not
> completely sold on it.
>
> When I code I personally find it very distracting when the minibuffer gets
> the focus :) Fortunately I believe this can easily be worked around, for
> example I'm pretty confident that somewhere someone has already written a
> Helm interface to list Skeletons.
>
> To recap: +1 for the work, although I would limit the use of skeletons for
> "file-level" boilerplate.
>
>
> On Wed, May 20, 2015 at 10:46 PM, Eric James Michael Ritz <
> notifications@github.com> wrote:
>
> > Personally I think it will be nice to have these in PHP Mode for
> > developers who use skeletons---useful reading material
> > <http://www.emacswiki.org/emacs/SkeletonMode>. I will add some
> > information to the README describing what they provide and some basic
> steps
> > for using them for any interested developers.
> >
> > But I will merge this at the end of the week at the earliest, because I
> > would like @syohex <https://github.com/syohex> and @trashofmasters
> > <https://github.com/trashofmasters> and anyone else interested to have a
> > chance to provide feedback before then.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/ejmr/php-mode/pull/250#issuecomment-104048540>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ejmr/php-mode/pull/250#issuecomment-104220902>.
>



-- 
David Arroyo Menéndez
http://www.davidam.com

**Comment by ejmr** [comment]
> ...but I am still unsure about having built-in PHP functions implemented as skeletons.

Just curious why.

**Comment by davidam** [comment]
skeletons is for templates, I did want make templates ...

2015-05-22 13:45 GMT+02:00 Eric James Michael Ritz <notifications@github.com
>:

> ...but I am still unsure about having built-in PHP functions implemented
> as skeletons.
>
> Just curious why.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ejmr/php-mode/pull/250#issuecomment-104640323>.
>



-- 
David Arroyo Menéndez
http://www.davidam.com

**Comment by ejmr** [comment]
Merged as 7eb69a8b584e1a5a49c1ecb927599705032535af.  Thanks for the patch David.  I updated the README to include you in the list of contributors.

**Comment by trashofmasters** [comment]
@ejmr: I think it's mostly down the fact there are a large number of skeletons that need be kept up-to-date manually. I really liked the way that your AutoYaSnippet package, which generated snippets on-demand. 

I'm very happy that @davidam code's made it in, and it shouldn't take too much effort to write a skeleton generator for builtin PHP function, should we ever need one.

**Comment by ejmr** [comment]
> ...and it shouldn't take too much effort to write a skeleton generator for builtin PHP function...

Interesting idea.  Perhaps we could use PHP's reflection mechanisms to automatically generate skeletons in the future.

**Comment by davidam** [comment]
Thanks for apply the patch!

2015-05-23 14:48 GMT+02:00 Eric James Michael Ritz <notifications@github.com
>:

> ...and it shouldn't take too much effort to write a skeleton generator for
> builtin PHP function...
>
> Interesting idea. Perhaps we could use PHP's reflection mechanisms to
> automatically generate skeletons in the future.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ejmr/php-mode/pull/250#issuecomment-104889469>.
>



-- 
David Arroyo Menéndez
http://www.davidam.com

**Comment by davidam** [comment]
I've continued working the idea is make a new git pull request?, perhaps a
git pull request per week is ok?

Thanks in advance

2015-05-23 22:36 GMT+02:00 David Arroyo Menéndez <davidam@gmail.com>:

> Thanks for apply the patch!
>
> 2015-05-23 14:48 GMT+02:00 Eric James Michael Ritz <
> notifications@github.com>:
>
>> ...and it shouldn't take too much effort to write a skeleton generator
>> for builtin PHP function...
>>
>> Interesting idea. Perhaps we could use PHP's reflection mechanisms to
>> automatically generate skeletons in the future.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/ejmr/php-mode/pull/250#issuecomment-104889469>.
>>
>
>
>
> --
> David Arroyo Menéndez
> http://www.davidam.com
>



-- 
David Arroyo Menéndez
http://www.davidam.com

**Comment by ejmr** [comment]
Feel free to make a pull request whenever you think it would be a useful contribution.

**Comment by trashofmasters** [comment]
@ejmr is there a way we can have skeletons to indent properly and according to the indentation style?
@davidam are planning to do more work on updating the skeletons to PHP 5?

**Comment by ejmr** [comment]
@trashofmasters I have no idea.  I actually didn't know skeletons were a part of Emacs proper until @davidam first submitted his pull request, heh.  He probably knows better than me, but I'll look into it too.

**Comment by davidam** [comment]
With skeleton you modify the indentation (see if in php-control-structures.el).

I've continued working with skeletons, perhaps could do another git
pull request with the wotk of
http://savannah.nongnu.org/projects/php-ext-el

Regards.

2015-06-19 1:36 GMT+02:00, Eric James Michael Ritz <notifications@github.com>:
> @trashofmasters I have no idea.  I actually didn't know skeletons were a
> part of Emacs proper until @davidam first submitted his pull request, heh.
> He probably knows better than me, but I'll look into it too.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/ejmr/php-mode/pull/250#issuecomment-113318916


-- 
David Arroyo Menéndez
http://www.davidam.com

**Comment by trashofmasters** [comment]
I had a look at the file `php-control-structures.el` and I see that the positioning of braces `{`, `}` is done inside a string and I fear it cannot be changed.

I'm not familiar with skeletons, so I don't know if they can be customized to follow the different coding standards used by php-mode. For example, if a skeleton defines:

```php
if ( condition ) { 
    _
}
```

then I suppose it's not possible to have the same skeleton also generate

```php
if ( condition )
{ 
    _
}
```

@davidam I'm probably wrong, do you think the above is already possible?

If not then I believe that we'll need to redefine each indentation-sensitive skeleton in every indentation style, for example:

- `php-mode/skeletons/symfony2/php-control-structures.el`
- `php-mode/skeletons/wordpress/php-control-structures.el`

@ejmr php-mode is now shipping with skeletons, but I noticed they need be loaded manually. I don't think we should enable them until we get the class skeleton to match with at least the PHP 5.0 syntax.

**Comment by davidam** [comment]
2015-06-21 13:00 GMT+02:00 Andrea Turso <notifications@github.com>:

> I had a look at the file php-control-structures.el and I see that the
> positioning of braces {, } is done inside a string and I fear it cannot
> be changed.
>
> I'm not familiar with skeletons, so I don't know if they can be customized
> to follow the different coding standards used by php-mode. For example, if
> a skeleton defines:
>
> if ( condition ) {     _}
>
> then I suppose it's not possible to have the same skeleton also generate
>
> if ( condition ){     _}
>
>  <https://github.com/davidam>
>

Yes.

> @davidam <https://github.com/davidam> I'm probably wrong, do you think
> the above is already possible?
>
> If not then I believe that we'll need to redefine each
> indentation-sensitive skeleton in every indentation style, for example:
>
>    - php-mode/skeletons/symfony2/php-control-structures.el
>    - php-mode/skeletons/wordpress/php-control-structures.el
>
>
This idea is better.


>
>
> @ejmr <https://github.com/ejmr> php-mode is now shipping with skeletons,
> but I noticed they need be loaded manually. I don't think we should enable
> them until we get the class skeleton to match with at least the PHP 5.0
> syntax.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ejmr/php-mode/pull/250#issuecomment-113885154>.
>



-- 
David Arroyo Menéndez
http://www.davidam.com
